### PR TITLE
Mbaas/storage network offline

### PR DESCRIPTION
**Pull request by MehrdadMzfr**
Things to consider before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [ ] Did you add unit tests?
* [x] Did you check UI tests on the sample app? They are not executed on CI.
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

Use separate HTTP clients for online and offline operations. This is to fail fast on online operations and still keep retrying on offline ones.

### CODE REVIEW

**Comment by nikitag-ms** [comment]
Guillaume suggested updating the changelog in the Android version. You can just copy it from there: https://github.com/microsoft/appcenter-sdk-android/pull/1258/files#diff-4ac32a78649ca5bdd8e0ba38b7006a1e.

**Comment by guperrot** [comment]
Any SDK behavior change or signature change has to be documented in the changelog.

File: CHANGELOG.md [hunk 1]
```diff
@@ -24,6 +24,7 @@ ___
 ### App Center Data
 
 * **[Breaking change]** Rename delegate method `data:didCompletePendingOperation:forDocument:withError:` from `MSRemoteOperationDelegate` to `data:didCompleteRemoteOperation:forDocumentMetadata:withError:`.
+* **[Fix]** Reduced retries on Data-related operations to fail fast and avoid the perception of calls "hanging".
```
Comment (achocron) [inline CHANGELOG.md]:
```suggestion
* **[Fix]** Reduce retries on Data-related operations to fail fast and avoid the perception of calls "hanging".
```

File: AppCenterData/AppCenterData/MSData.m [hunk 2]
```diff
@@ -661,6 +664,7 @@ - (void)deleteFromCosmosDbWithPartition:(NSString *)partition
   [self performCosmosDbOperationWithPartition:partition
                                    documentId:documentId
                                    httpMethod:kMSHttpMethodDelete
+                                   httpClient:self.httpClientWithRetrier
```
Comment (nikitag-ms) [inline AppCenterData/AppCenterData/MSData.m]: This one isn't processing a pending delete, right? Shouldn't this one be without a retryer?

File: AppCenterData/AppCenterData/MSData.m [hunk 3]
```diff
@@ -556,6 +557,7 @@ - (void)upsertFromCosmosDbWithPartition:(NSString *)partition
       performCosmosDbOperationWithPartition:partition
                                  documentId:documentId
                                  httpMethod:kMSHttpMethodPost
+                                 httpClient:self.httpClientWithRetrier
```
Comment (nikitag-ms) [inline AppCenterData/AppCenterData/MSData.m]: I might be reading this incorrectly, but this seems like an online operation (not a pending one). If so, then shouldn't this be the no retrier instance?

File: AppCenterData/AppCenterData/MSData.m (hunk 2 above)
Comment (MehrdadMzfr) [reply AppCenterData/AppCenterData/MSData.m]: this is actually coming from `processPendingOperations` call. for regular delete it uses:
```- (void)deleteDocumentWithID:(NSString *)documentID
                   partition:(NSString *)partition
                writeOptions:(MSWriteOptions *_Nullable)writeOptions
           completionHandler:(MSDocumentWrapperCompletionHandler)completionHandler {```

File: AppCenterData/AppCenterData/MSData.m (hunk 3 above)
Comment (MehrdadMzfr) [reply AppCenterData/AppCenterData/MSData.m]: good point, refactored the code, this method can be called from both CreateOrReplace and ProcessPendingOperations. therefore, it's required to path the desired httpClient as a parameter.

File: CHANGELOG.md [hunk 4]
```diff
@@ -24,6 +24,7 @@ ___
 ### App Center Data
 
 * **[Breaking change]** Rename delegate method `data:didCompletePendingOperation:forDocument:withError:` from `MSRemoteOperationDelegate` to `data:didCompleteRemoteOperation:forDocumentMetadata:withError:`.
+* **[Fix]** Reduce retries on Data-related operations to fail fast and avoid the perception of calls "hanging".
```
Comment (jaelim-ms) [inline CHANGELOG.md]: We've already released 2.4.0. Please move this to a newer version.

**Pull request merged**